corehq/motech/openmrs/finders.py

```diff
@@ -0,0 +1,166 @@
+
+
+# JsonpathValueMap is for comparing OpenMRS patients with CommCare
+# cases.
+#
+# The `jsonpath` attribute is used for retrieving values from an
+# OpenMRS patient and the `value_map` attribute is for converting
+# OpenMRS concept UUIDs to CommCare property values, if necessary.
+JsonpathValuemap = namedtuple('JsonpathValuemap', ['jsonpath', 'value_map'])
+
+
+def get_caseproperty_jsonpathvaluemap(jsonpath, value_source):
+    """
+    Used for updating property_map to map case properties to OpenMRS
+    patient property-, attribute- and concept values.
+
+    :param jsonpath: The path to a value in an OpenMRS patient JSON object
+    :param value_source: A case_config ValueSource instance
+    :return: A single-item dictionary with the name of the case
+             property as key, and a JsonpathValuemap as value. If
+             value_source is a constant, then there is no corresponding
+             case property, so the function returns an empty dictionary
+    """
+    if value_source['doc_type'] == 'ConstantString':
+        return {}
+    if value_source['doc_type'] == 'CaseProperty':
+        return {value_source['case_property']: JsonpathValuemap(jsonpath, {})}
+    if value_source['doc_type'] == 'CasePropertyConcept':
+        value_map = {v: k for k, v in value_source['value_concepts'].items()}
+        return {value_source['case_property']: JsonpathValuemap(jsonpath, value_map)}
+
+
+class WeightedPropertyPatientFinder(PatientFinderBase):
+    """
+    Finds patients that match cases by assigning weights to matching
+    property values, and adding those weights to calculate a confidence
+    score.
+    """
+    def __init__(self, searchable_properties, property_weights, threshold=1, confidence_margin=0.1):
+        """
+        Initialise the instance
+
+        :param searchable_properties: Properties that can be used to
+               search for patients in OpenMRS
+        :param property_weights: A dictionary of case property names
+               and their weights
+        :param threshold: The sum of weights that must be met in order
+               for a patient to be considered a match
+        """
+        self.searchable_properties = searchable_properties
+        self.property_weights = property_weights
+        self.threshold = threshold
+        self.confidence_margin = confidence_margin
+        self.property_map = {}
+
+    def set_property_map(self, case_config):
+        """
+        Set self.property_map to map OpenMRS properties and attributes
+        to case properties.
+        """
+        # Example value of case_config::
+        #
+        #     {
+        #       "person_properties": {
+        #         "birthdate": {
+        #           "doc_type": "CaseProperty",
+        #           "case_property": "dob"
+        #         }
+        #       },
+        #       // ...
+        #     }
+        #
+
+        for person_prop, value_source in case_config['person_properties'].items():
```

Comment:
Should we use `get` here `case_config['person_properties']` instead to avoid failure in case the key is missing? Similarly for other 3 `property_map` iteration after this..

Reply:
`case_config` is an instance of [OpenmrsCaseConfig](https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/openmrs_config.py#L90) so it will always have `person_properties` ... but you make a good point, because that is not obvious. Maybe it would be better to use it as an attribute, `case_config.person_properties`, instead of a key.